### ISSUE TITLE
EC2 IAM thread vanish

### ISSUE THREAD

**Issue by jeromerobert** [archiecobbs/s3backer#115]
After while my s3backer folder freeze. syslog shows many 400 error. `tcpdump` says that the IAM token as expired. This only happen in fork mode, not in foreground (-d flag). This happen since I upgraded from Debian Stretch to Debian Buster. I'm using the `--accessEC2IAM` option.

Using `gdb --pid` I find that the EC2 IAM thread is no long existing after the fork. My (still very partial) understanding is that it's created in the parent process while it should be created after the fork.

**Comment by jeromerobert** [comment]
I guess this is similar to #112. I really don't understand why it was working before, because the EC2 IAM thread has always been created before calling `fuse_main` (i.e forking).

**Comment by archiecobbs** [comment]
Probably #112. Please try the latest code version and see if the problem is fixed.

**Comment by jeromerobert** [comment]
No #112 is about the block cache thread, not the IAM token thread. I'm already using the latest code version: da27ae6.

**Comment by archiecobbs** [comment]
Yes, you're right. The previous fix for creating threads neglected to handle the IAM updater thread.

Please try this patch:

```
diff --git a/http_io.c b/http_io.c
index 363d21b..1b54698 100644
--- a/http_io.c
+++ b/http_io.c
@@ -413,13 +413,9 @@ http_io_create(struct http_io_conf *config)
     /* Initialize cURL */
     curl_global_init(CURL_GLOBAL_ALL);
 
-    /* Initialize IAM credentials and start updater thread */
-    if (config->ec2iam_role != NULL) {
-        if ((r = update_iam_credentials(priv)) != 0)
-            goto fail5;
-        if ((r = pthread_create(&priv->iam_thread, NULL, update_iam_credentials_main, priv)) != 0)
-            goto fail5;
-    }
+    /* Initialize IAM credentials */
+    if (config->ec2iam_role != NULL && (r = update_iam_credentials(priv)) != 0)
+        goto fail5;
 
     /* Take ownership of non-zero block bitmap */
     priv->non_zero = config->nonzero_bitmap;
@@ -881,7 +877,13 @@ http_io_block_hash_prefix(s3b_block_t block_num)
 static int
 http_io_create_threads(struct s3backer_store *s3b)
 {
-    return 0;
+    struct http_io_private *const priv = s3b->data;
+    struct http_io_conf *const config = priv->config;
+
+    /* Start IAM updater thread */
+    if (config->ec2iam_role == NULL)
+        return 0;
+    return pthread_create(&priv->iam_thread, NULL, update_iam_credentials_main, priv);
 }
 
 static int

```

**Comment by archiecobbs** [comment]
Should be fixed by 81860c9. Please re-open if not.

**Comment by abinnie** [comment]
`http_io_destroy` always tries to kill the IAM thread if a role is provided, even before fork. This leads to:
```
# s3backer [...]
s3backer: auto-detecting block size and total file size...
Segmentation fault
```
This patch seems to fix the issue:
```
diff --git a/http_io.c b/http_io.c
index 14c2fe4..9694136 100644
--- a/http_io.c
+++ b/http_io.c
@@ -158,6 +158,7 @@ struct http_io_private {
     u_int                       *non_zero;      // config->nonzero_bitmap is moved to here
     pthread_t                   iam_thread;     // IAM credentials refresh thread
     u_char                      shutting_down;
+    u_char                      iam_thread_created;
 
     /* Encryption info */
     const EVP_CIPHER            *cipher;
@@ -464,7 +465,7 @@ http_io_destroy(struct s3backer_store *const s3b)
 
     /* Shut down IAM thread */
     priv->shutting_down = 1;
-    if (config->ec2iam_role != NULL) {
+    if (priv->iam_thread_created) {
         (*config->log)(LOG_DEBUG, "waiting for EC2 IAM thread to shutdown");
         if ((r = pthread_cancel(priv->iam_thread)) != 0)
             (*config->log)(LOG_ERR, "pthread_cancel: %s", strerror(r));
@@ -882,9 +883,11 @@ http_io_create_threads(struct s3backer_store *s3b)
     int r;
 
     /* Start IAM updater thread if appropriate */
-    if (config->ec2iam_role != NULL
-      && (r = pthread_create(&priv->iam_thread, NULL, update_iam_credentials_main, priv)) != 0)
-        return r;
+    if (config->ec2iam_role != NULL) {
+        if ((r = pthread_create(&priv->iam_thread, NULL, update_iam_credentials_main, priv)) != 0)
+            return r;
+        priv->iam_thread_created = 1;
+    }
 
     /* Done */
     return 0;

```

**Comment by archiecobbs** [comment]
D'oh! Thanks. Should be fixed in 3cfdfc4.

**Comment by archiecobbs** [comment]
Just released 1.5.4 two days ago.
```
Version 1.5.4 released October 8, 2019

    - Only set "x-amz-server-side-encryption" header with PUT requests (issue #116)
    - Don't kill IAM thread unless actually started (issue #115).

Version 1.5.3 released August 9, 2019

    - Fixed bug where IAM update thread was killed after fork (issue #115)
    - Fixed use-after-free bug in block_cache_verified() (issue #113)
    - Fixed use-after-free bug when updating IAM credentials (pr #114)
    - Fixed bug in test mode that was causing bogus I/O errors
```

Also, tags for 1.5.3 and 1.5.4 pushed now (forgot).
